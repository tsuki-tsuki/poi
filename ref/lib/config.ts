/* eslint-disable @typescript-eslint/no-explicit-any */
import { set, get, isEqual, keys } from 'lodash'
import EventEmitter from 'events'
import CSON from 'cson'
import fs from 'fs-extra'
import path from 'path'
import dbg from './debug'
import defaultConfig from './default-config'
import { mergeConfig, warn } from './utils'

const { EXROOT } = global
const configPath = path.join(EXROOT, 'config.cson')

const DEFAULT_CONFIG_PATH_REGEXP = new RegExp(`^[${keys(defaultConfig).join('|')}]`)

export type Config = { [key: string]: Config }

class PoiConfig extends EventEmitter {
  configData: Config
  defaultConfigData: typeof defaultConfig

  constructor() {
    super()
    this.configData = {}
    try {
      fs.accessSync(configPath, fs.constants.R_OK | fs.constants.W_OK)
      this.configData = mergeConfig(defaultConfig, CSON.parseCSONFile(configPath)) as Config
      dbg.log(`Config loaded from: ${configPath}`)
    } catch (e) {
      dbg.log(e)
    }
    this.defaultConfigData = defaultConfig
  }

  /**
   * get a config value at give path
   * @param {String | String[]} path the given config location
   * @param value value to fallback if queried config is undefined
   */
  get = (path: string | string[] = '', value?: any) => {
    if (path === '') {
      return this.configData
    }
    if (dbg.isEnabled()) {
      const stringPath = Array.isArray(path) ? path.join('.') : path
      if (
        DEFAULT_CONFIG_PATH_REGEXP.test(stringPath) &&
        value !== undefined &&
        !isEqual(get(this.defaultConfigData, path), value)
      ) {
        warn('There might be a mssing config default, check', stringPath, value)
      }
    }
    return get(this.configData, path, this.getDefault(path, value))
  }

  /**
   * get default config value at give path
   * @param {String | String[]} path the given config location
   */
  getDefault = (path: string | string[] = '', value?: any) => {
    if (path === '') {
      return this.defaultConfigData
    }
    return get(this.defaultConfigData, path, value)
  }

  /**
   * set a config value at give path
   * @param {String | String[]} path the given config location
   * @param value value to overwrite, if the path belongs to poi's default config, will reset to default value
   */
  set = (path: string | string[], value: any) => {
    if (get(this.configData, path) === value) {
      return
    }
    if (value === undefined && this.getDefault(path) !== undefined) {
      value = this.getDefault(path)
    }
    set(this.configData, path, value)
    path = Array.isArray(path) ? path.join('.') : path
    this.emit('config.set', path, value)
    this.save()
  }

  /**
   * set a config value only when it is not set (addition only)
   * @param {String | String[]} path the given config location
   * @param value value to overwrite, leaving undefined will remove the config
   */
  setDefault = (path: string | string[], value: any) => {
    if (this.get(path) === undefined) {
      this.set(path, value)
    }
  }

  /**
   * save current config to file
   */
  save = () => {
    try {
      fs.writeFileSync(configPath, CSON.stringify(this.configData, undefined, 2))
    } catch (e) {
      console.warn(e)
    }
  }

  /**
   * remove a config at given path
   * @param {String | String[]} path path to remove
   */
  delete = (path: string | string[]) => {
    if (typeof this.get(path) !== 'undefined') {
      let p = this.configData
      const subpath = Array.isArray(path) ? path : path.split('.')
      for (const sub of subpath.slice(0, subpath.length - 1)) {
        p = p[sub]
      }
      delete p[subpath[subpath.length - 1]]
    }
  }
}

const config = new PoiConfig()
config.setMaxListeners(100)

export type ConfigInstance = typeof config

export default config
