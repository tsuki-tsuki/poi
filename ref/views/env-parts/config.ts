import * as remote from '@electron/remote'
import type { ConfigInstance } from 'lib/config'
export type { ConfigInstance, Config } from 'lib/config'

export const config: ConfigInstance = remote.require('./lib/config')

// @ts-expect-error backward compatibility
window.config = config
