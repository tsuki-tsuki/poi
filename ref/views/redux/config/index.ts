import { cloneDeep } from 'lodash'
import { type Config, config } from 'views/env-parts/config'

import { reduxSet } from 'views/utils/tools'
import { createConfigAction } from '../actions'
import { createSlice } from '@reduxjs/toolkit'

const configSlice = createSlice({
  name: 'config',
  initialState: cloneDeep(config.get('')) as Config,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(createConfigAction, (state, { payload }) => {
      const { path, value } = payload
      return reduxSet(state, path.split('.'), value)
    })
  },
})

export const reducer = configSlice.reducer
