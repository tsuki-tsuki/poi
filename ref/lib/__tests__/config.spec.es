jest.mock('fs-extra', () => ({
  accessSync: jest.fn(),
  writeFileSync: jest.fn(),
  constants: {},
}))
jest.mock('cson')
jest.mock('../default-config.ts', () => ({ foo: 'bar' }))
import config from '../config'
import fs from 'fs-extra'
import CSON from 'cson'

describe('config', () => {
  it('initially, should be filled with default value', () => {
    expect(CSON.parseCSONFile.mock.calls).toMatchSnapshot()
    expect(config.get('', null)).toMatchSnapshot()
  })

  it('should be set and get a string correctly', () => {
    config.set('path.to.value', 'Dead Beef')
    expect(config.get('path.to.value', null)).toBe('Dead Beef')
    expect(fs.writeFileSync).toHaveBeenCalled()
    expect(CSON.stringify.mock.calls).toMatchSnapshot()
    expect(fs.writeFileSync.mock.calls[0][0].includes(global.EXROOT)).toBe(true)
    expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot()
  })
  it('should be set and get a object correctly', () => {
    config.set('another', {
      a: 'b',
      c: 0xbb,
    })
    expect(fs.writeFileSync).toHaveBeenCalled()
    expect(CSON.stringify.mock.calls).toMatchSnapshot()
    expect(fs.writeFileSync.mock.calls[0][0].includes(global.EXROOT)).toBe(true)
    expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot()
    expect(config.get('another', null)).toEqual({
      a: 'b',
      c: 0xbb,
    })
  })
  it('should be get a null', () => {
    expect(config.get('null.path', null)).toBe(null)
  })
})
