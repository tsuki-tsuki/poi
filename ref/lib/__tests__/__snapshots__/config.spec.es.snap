// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config initially, should be filled with default value 1`] = `Array []`;

exports[`config initially, should be filled with default value 2`] = `
Object {
  "foo": "bar",
}
`;

exports[`config should be set and get a object correctly 1`] = `
Array [
  Array [
    Object {
      "another": Object {
        "a": "b",
        "c": 187,
      },
      "foo": "bar",
      "path": Object {
        "to": Object {
          "value": "Dead Beef",
        },
      },
    },
    undefined,
    2,
  ],
]
`;

exports[`config should be set and get a object correctly 2`] = `"CSON Stringified"`;

exports[`config should be set and get a string correctly 1`] = `
Array [
  Array [
    Object {
      "foo": "bar",
      "path": Object {
        "to": Object {
          "value": "Dead Beef",
        },
      },
    },
    undefined,
    2,
  ],
]
`;

exports[`config should be set and get a string correctly 2`] = `"CSON Stringified"`;
